### PR TITLE
Write parser and encoder in pure functions

### LINKED ISSUE #31: Functional streams
Streams are nice but introduce an overhead for sync APIs would be great to rewrite the streams paser/stringifyer that way:

```
function parseSVGPathData(stateObject, chunk) {

  // Parsing

  return [
    newStateObject,
    commands
  ];
}
```

This would reduce this to a single function call:
https://github.com/nfroidure/svg-pathdata/blob/master/src/SVGPathData.js#L122-L136

And allow to create non-NodeJS stream (browser streams for example).

### PR DESCRIPTION
**Pull request by pioug**
Fixes #31 and #33 

Opening the PR for visibility. But it is still ongoing
- [x] Write pure functions
- [x] Run tests in latest Node on CI
- [ ] Update documentation

### Proposed changes
- Fix TS errors during build 
- Expose synchronous API for parser and encoder

### Code quality
- [x] I made some tests for my changes
- [ ] I added my name in the
 [contributors](https://docs.npmjs.com/files/package.json#people-fields-author-contributors)
 field of the package.json file

### License
To get your contribution merged, you must check the following.

- [x] I read the project license in the LICENSE file
- [x] I agree with publishing under this project license

### Join
- [x] I wish to join the core team
- [x] I agree that with great powers comes responsibilities
- [x] I'm a nice person

My NPM username: pioug

### CODE REVIEW

File: tests/encoder.mocha.js [hunk 1]
```diff
@@ -6,17 +6,15 @@ const { SVGPathDataEncoder } = require('..');
 
 describe('SVGPathDataEncoder', () => {
 
-  it('should still work when the new operator is forgotten', () => {
-    assert.doesNotThrow(() => {
+  it('should not work when the command is forgotten', () => {
+    assert.throws(() => {
       new SVGPathDataEncoder();
```
Comment (NaridaL) [inline tests/encoder.mocha.js]: as SVGPathDataEncoder is now a pure function, drop the `new` operator.

File: README.md [hunk 2]
```diff
@@ -55,26 +55,18 @@ console.log(pathData.commands);
...
-parser.write('   ');
-parser.write('M 10');
-parser.write(' 10');
-// {type: SVGPathData.MOVE_TO, relative: false, x: 10, y: 10 }
+let parser = new SVGPathDataParser('H 60');
+// [{ type: SVGPathData.HORIZ_LINE_TO, relative: false, x: 60 }]
 
-parser.write('H 60');
-// {type: SVGPathData.HORIZ_LINE_TO, relative: false, x: 60 }
+let parser = new SVGPathDataParser('V 60');
```
Comment (NaridaL) [inline README.md]: No new op here either.

**Comment by NaridaL** [comment]
Hey @pioug thanks for the work, looks mostly good so far. Unfortunately, the "streaming" part, as exemplified by the following example from the readme:

```
parser.write('   ');
parser.write('M 10');
parser.write(' 10');
// {type: SVGPathData.MOVE_TO, relative: false, x: 10, y: 10 }
```

(i.e. you can pass strings in chunks) is no longer possible.

Changing it to fix that shouldn't be too difficult, basically you want to to model the parser state `type ParserState = {curNumber: string, curCommand: Command, state: number}` as an explicit object, which you optionally accept as a parameter in the pure function and return in addition to the parsed commands so the user can continue parsing with the current parser state.

To cite @nfroidure's original issue:

```
function parseSVGPathData(stateObject: ParserState = START_PARSER_STATE, chunk: string) {
  // Parsing

  return [
    newStateObject,
    commands
  ];
}
```

**Comment by NaridaL** [comment]
Alternatively, instead of going the pure function way, you could just implement it the way the transformers currently are (with closures):

```
function createParser() {
  let curState =
  let curNumber =
  let curCommand = 
  return function parseChunk(chunk: string, commands: Command[] = []) {
    commands.push(...)
    return commands
  }
}
```

This would have the advantage of consistency. Rewriting all the transformers as pure functions seems like  a lot of work. (Although if you decide to just have the parser as a pure function and leave the transformers as they are I'll still merge it...)

**Comment by pioug** [comment]
@NaridaL Thanks a lot for reviewing 🙏 I will find some time to update but feel free to commit any meanwhile!

**Comment by NaridaL** [comment]
@pioug I finished the PR and committed it.  Thanks a lot for your help!

**Pull request merged**